Provision ipaserver in ca-less mode for WebUI tests.

miskopo opened the pull request:
Templates and ansible playbooks for ca-less provisioning of ipa server
used for WebUI testing. Certificates are generated locally and provided
to the installer.

Signed-off-by: Michal Polovka <mpolovka@redhat.com>

miskopo commented:
I'd also like to generate `passwd_file`, `noise_file` and `SKI` randomly, however, I've not found suitable ansible option.

flo-renaud reviewed (commented):
@miskopo 
there is a lot of code that could probably be shared instead of being duplicated. You can have a look at RunPytest for instance in tasks/task.py, that accepts an additional parameter "selinux_enforcing" and you can make something similar with RunWebuiTests and a parameter "caless".
This way, we could define the jobs with:
```
  fedora-latest/test_webui_xxx:
    requires: [fedora-latest/build]
    priority: 50
    job:
      class: RunWebuiTests
      args:
        build_url: '{fedora-latest/build_url}'
        caless: True
        test_suite: test_webui/test_xxx.py
        template: *ci-master-latest
        timeout: 2400
        topology: *ipaserver
```
When RunWebuiTests is called with caless: True, it would use the same Vagrant template Vagrantfile.ipaserver but provide the parameter and call the playbook provision_ipaserver.yml with this extra param (see webui.vars.yaml). This way, you just need to modify the existing playbook to handle cafull/caless instead of copy-pasting the whole playbook.

miskopo commented:
Hello @flo-renaud , thank you for your feedback. I've implemented requested changes, however I am not sure I correctly understood `caless` variable propagation in `tasks.py`.

flo-renaud commented on ansible/roles/machine/provision_ipaserver/tasks/main.yml:
```diff
@@ -5,6 +5,102 @@
+- name: fetch hostname
+  command: hostname -f
+  register: master_hostname
+  when: caless == true
```
The hostname is set earlier to` {{ inventory_hostname }}.ipa.test` (in https://github.com/freeipa/freeipa-pr-ci/blob/master/ansible/roles/machine/provision/tasks/main.yml#L95). You can remove this task and use `{{ inventory_hostname }}.ipa.test` instead of `{{ master_hostname }}` later on.

flo-renaud reviewed (commented):
Hi @miskopo 
please find some inline comments.
If you want to test the playbook, you can create a temp directory in your freeipa-pr-ci/jobs repo, and inside this directory create a Vagrantfile and vars.yml similar to the ones that PRCI generates. See for instance in http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/425cf742-7af1-11eb-a025-fa163e5f891e/
Then launch `vagrant up` from inside this directory, you will be able to see the playbook output and errors.

flo-renaud commented on ansible/roles/machine/provision_ipaserver/tasks/main.yml:
```diff
@@ -14,7 +110,13 @@
     -a Secret.123
     --setup-dns
     --setup-kra
```
--setup-kra cannot be used in ca-less mode

flo-renaud reviewed (commented):
An additional comment: the `lookup('password' ...)` call generates the file on the control node, not on the managed node. You may want to use `dd` to generate random chars, for instance:
```
 - name: create noise file
   shell: dd count=10 bs=1024 if=/dev/random of="/tmp/noise_file"
   when: caless == true
```

miskopo commented:
Hello @flo-renaud ,

I've implemented the modifications you had pointed out and tried to test it. All went well until step `TASK [machine/provision_ipaserver : install freeipa in  caless mode] `, which failed with rather cryptic error 

```
fatal: [master]: FAILED! => {
"changed": true, 
"cmd": "ipa-server-install -U -n ipa.test -r IPA.TEST -p Secret.123 -a Secret.123 --setup-dns --http-cert-file=/tmp/nssdb/server.p12 --dirsrv-cert-file=/tmp/nssdb/server.p12 --http-pin Secret123 --dirsrv-pin Secret123 --forwarder=#IP_HIDDEN --no-pkinit\n",
 "delta": "0:00:02.681680",
 "end": "2021-03-06 18:46:13.879974", 
"msg": "non-zero return code", 
"rc": 1, 
"start": "2021-03-06 18:46:11.198294",
 "stderr": "Attempted \"__iter__\" operation on ASN.1 schema object
The ipa-server-install command failed. See /var/log/ipaserver-install.log for more information", 
"stderr_lines": ["Attempted \"__iter__\" operation on ASN.1 schema object", "The ipa-server-install command failed. See /var/log/ipaserver-install.log for more information"], "stdout": "\nThe log file for this installation can be found in /var/log/ipaserver-install.log
==============================================================================
This program will set up the IPA Server.
Version 4.10.0.dev

This includes:
  * Configure the NTP client (chronyd)
  * Create and configure an instance of Directory Server 
  * Create and configure a Kerberos Key Distribution Center (KDC)
  * Configure Apache (httpd)
  * Configure DNS (bind)
   
  * Warning: skipping DNS resolution of host master.ipa.test", 
  "stdout_lines": ["", "The log file for this installation can be found in /var/log/ipaserver-install.log", "==============================================================================", "This program will set up the IPA Server.", "Version 4.10.0.dev", "", "This includes:", "  * Configure the NTP client (chronyd)", "  * Create and configure an instance of Directory Server", "  * Create and configure a Kerberos Key Distribution Center (KDC)", "  * Configure Apache (httpd)", "  * Configure DNS (bind)", "", "Warning: skipping DNS resolution of host master.ipa.test"]}
```

flo-renaud commented:
> ```
>  "stderr": "Attempted \"__iter__\" operation on ASN.1 schema object
> ```

There is more information in the server install log. You can do `vagrant ssh` in order to log into the machine and then have a look at the log file.
The error happens because the server certificate does not contain a Subject Alternative Name extension with the DNS name. The certutil cmd generating the server cert needs to use `-8 {{ inventory_hostname }}.ipa.test`.

miskopo commented:
@flo-renaud Thank you very much, all issues are now resolved and playbook works as intended. Full execution log follows:

```
PLAY [all] *********************************************************************

TASK [Gathering Facts] *********************************************************
ok: [master]

TASK [machine/provision : get DNS server from resolv.conf (fedora <= 32)] ******
skipping: [master]

TASK [machine/provision : get DNS server from resolvectl (fedora >= f33)] ******
changed: [master]

TASK [machine/provision : set dns forwarder fact (fedora <= 32)] ***************
skipping: [master]

TASK [machine/provision : set dns forwarder fact (fedora >= f33)] **************
ok: [master]

TASK [machine/provision : ensure file already exists at template dest to work around 'invalid selinux context' issue] ***
changed: [master]

TASK [machine/provision : create test config file] *****************************
changed: [master]

TASK [machine/provision : add PR build repository] *****************************
changed: [master]

TASK [machine/provision : update packages] *************************************
skipping: [master]

TASK [machine/provision : install freeipa packages] ****************************
changed: [master]

TASK [machine/provision : install client packages] *****************************
skipping: [master]

TASK [machine/provision : install Fedora 27 specific tests dependencies] *******
skipping: [master]

TASK [machine/provision : create directory to save installed packages logs] ****
ok: [master]

TASK [machine/provision : get all packages] ************************************
[WARNING]: Consider using the yum, dnf or zypper module rather than running
'rpm'.  If you need to use command because yum, dnf or zypper is insufficient
you can add 'warn: false' to this command task or set 'command_warnings=False'
in ansible.cfg to get rid of this message.
changed: [master]

TASK [machine/provision : create hosts file from template] *********************
changed: [master]

TASK [machine/provision : set hostname] ****************************************
changed: [master]

TASK [machine/provision : set selinux to enforcing] ****************************
skipping: [master]

TASK [include_role : utils] ****************************************************

TASK [utils : write swap file] *************************************************
changed: [master]

TASK [utils : set swap file permissions] ***************************************
changed: [master]

TASK [utils : create swap file] ************************************************
changed: [master]

TASK [utils : enable swapfile] *************************************************
changed: [master]

TASK [utils : add swapfile to /etc/fstab] **************************************
changed: [master]

TASK [machine/provision_ipaserver : deploy script to execute webui tests] ******
changed: [master]

TASK [machine/provision_ipaserver : ensure nss-tools is present] ***************
changed: [master]

TASK [machine/provision_ipaserver : create nssdb directory] ********************
changed: [master]

TASK [machine/provision_ipaserver : create passwd file] ************************
changed: [master]

TASK [machine/provision_ipaserver : create noise file] *************************
changed: [master]

TASK [machine/provision_ipaserver : generate SKID] *****************************
ok: [master]

TASK [machine/provision_ipaserver : create certificate and keys database] ******
changed: [master]

TASK [machine/provision_ipaserver : create self-signed ca cert] ****************
changed: [master]

TASK [machine/provision_ipaserver : create server certificate] *****************
changed: [master]

TASK [machine/provision_ipaserver : export server certificate] *****************
changed: [master]

TASK [machine/provision_ipaserver : install freeipa in caless mode] ************
changed: [master]

TASK [machine/provision_ipaserver : install freeipa] ***************************
skipping: [master]

TASK [machine/provision_ipaserver : set IPA to development mode] ***************
changed: [master]

TASK [machine/provision_ipaserver : restart httpd] *****************************
changed: [master]

TASK [machine/provision_ipaserver : create .ipa dir] ***************************
changed: [master]

TASK [machine/provision_ipaserver : copy default.conf and other required files] ***
changed: [master]

TASK [machine/provision_ipaserver : set wait_for_dns=5] ************************
changed: [master]

TASK [machine/provision_ipaserver : create ~/.ipa/.dmpw] ***********************
changed: [master]

TASK [machine/provision_ipaserver : create ui_test.conf] ***********************
changed: [master]

TASK [machine/provision_ipaserver : add service csr] ***************************
changed: [master]

TASK [machine/provision_ipaserver : add host csr] ******************************
changed: [master]

TASK [machine/provision_ipaserver : add user csr] ******************************
changed: [master]

PLAY RECAP *********************************************************************
master                     : ok=37   changed=33   unreachable=0    failed=0    skipped=7    rescued=0    ignored=0  
```

flo-renaud commented:
@miskopo the code LGTM but I only tested the `vagrant up` part, not the part executed when a PR is raised for a webui test. I know that @wladich has a personal test runner, maybe he can explain to you how to setup your own prci runner and you can create a PR against your freeipa fork, with a temp commit launching one webui test in caless mode and one webui test in cafull mode.

A potential improvement would be to create the pkinit cert, but that's not a hard requirement for the moment.

wladich commented:
The proposed patch does  not work for me. I have tested it in https://github.com/wladich/freeipa/pull/44: I have added `caless: yes` parameter for RunWebuiTests but [the runner log](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/9bafcf12-8105-11eb-b7a9-fa163ebfa111/runner.log.gz) shows it was not used:
```
2021-03-09 18:32:55,573    DEBUG  TASK [machine/provision_ipaserver : install freeipa in caless mode] ************
2021-03-09 18:32:55,618    DEBUG  skipping: [master]
```

flo-renaud commented:
@miskopo you will need to override the _before() method of the RunWebuiTests class and make it create vars.yml with a value for caless.

wladich reviewed (commented):
I am not an expert in certificates so could you please tell, do we really need this complicated way of generating the certificate with manual creation of noise file and interactive execution of certutil? (I really do not like those interactive shells).

Are there any special requirements which stop us from using openssl? Something like
```
openssl req -x509 -newkey rsa:4096 -keyout key.pem -out cert.pem -days 365 -nodes -subj '/CN=testrelm.test
openssl pkcs12 -export -out Cert.p12 -in cert.pem -inkey key.pem  -passout pass:root
```

Reference: 
* https://stackoverflow.com/a/10176685
* https://help.hcltechsw.com/appscan/Standard/9.0.3/en-US/t_ConvertthepfxCertificatetopemFormat068.html

miskopo commented on ansible/roles/machine/provision_ipaserver/tasks/main.yml:
```diff
@@ -5,6 +5,95 @@
+      path: /tmp/nssdb
+      state: directory
+
+  - name: create passwd file
```
I am trying this approach, but I have difficulties getting it to work

miskopo commented:
> I am not an expert in certificates so could you please tell, do we really need this complicated way of generating the certificate with manual creation of noise file and interactive execution of certutil? (I really do not like those interactive shells).
> 
> Are there any special requirements which stop us from using openssl? Something like
> 
> ```
> openssl req -x509 -newkey rsa:4096 -keyout key.pem -out cert.pem -days 365 -nodes -subj '/CN=testrelm.test
> openssl pkcs12 -export -out Cert.p12 -in cert.pem -inkey key.pem  -passout pass:root
> ```
> 
> Reference:
> 
>     * https://stackoverflow.com/a/10176685
> 
>     * https://help.hcltechsw.com/appscan/Standard/9.0.3/en-US/t_ConvertthepfxCertificatetopemFormat068.html

Simply, in other `caless` tests in `freeipa`, the certificates are being generated using `certutil`. I've tried this approach, i.e. using `openssl`, but couldn't get `ipa` to accept the certificate. If you wish, we can work on it together, as of now, I am leaving it as it is.

miskopo commented:
@wladich @flo-renaud Thank you for your feedback, I've implemented requested changes.

flo-renaud commented on ansible/roles/machine/provision_ipaserver/tasks/main.yml:
```diff
@@ -5,6 +5,111 @@
+        key_usage_critical: yes
+        subject:
+          commonName: CA
+          organizationName: "{{ pki_org }}"
```
With these settings the CA subject is `CN = CA, CN = CA, O = Example Org`

miskopo replied:
Hello Flo! Thank you for your feedback. I am not quite sure here what's the desired subject and how to achieve it.

flo-renaud commented on ansible/roles/machine/provision_ipaserver/tasks/main.yml:
```diff
@@ -5,6 +5,111 @@
+        common_name: "{{ meta_fqdn }}"
+        subject:
+          commonName: "{{ meta_fqdn }}"
+          organizationName: "{{ pki_org }}"
```
With these settings the subject is `CN = master.ipa.test, CN = master.ipa.test, O = Example Org`

flo-renaud commented on ansible/roles/machine/provision_ipaserver/tasks/main.yml:
```diff
@@ -5,6 +5,111 @@
+          commonName: "{{ meta_fqdn }}"
+          organizationName: "{{ pki_org }}"
+      with_items:
+        - "{{ pki_host }}"
```
Why need `with_items` if there is a single item? You can replace `{{ item }}` in the above lines with `{{ pki_host }}`

flo-renaud commented on ansible/roles/machine/provision_ipaserver/tasks/main.yml:
```diff
@@ -5,6 +5,111 @@
+        other_certificates: "{{ pki_dir }}/ca1.crt"
+        state: present
+      with_items:
+        - "{{ pki_host }}"
```
Why use `with_items` if there is a single item?

flo-renaud commented on ansible/roles/machine/provision_ipaserver/tasks/main.yml:
```diff
@@ -5,6 +5,111 @@
+      openssl_pkcs12:
+        action: export
+        path: "{{ pki_dir }}/ca1/{{ item }}.p12"
+        friendly_name: "{{ item }}"
```
We usually name the cert 'Server-Cert', not 'master'.

flo-renaud commented on ansible/roles/machine/provision_ipaserver/templates/extensions.cnf.j2:
```diff
@@ -0,0 +1,24 @@
+[ca_cert]
```
I don't see this file used.

flo-renaud reviewed (commented):
Hi @miskopo 
please find inline comments.

miskopo commented:
@wladich hello! Could you please try this solution on your runner? Thank you :)

wladich commented on ansible/roles/machine/provision_ipaserver/tasks/main.yml:
```diff
@@ -5,6 +5,105 @@
 
+- block:
+    - name: set PKI related variables
+      set_fact:
```
`set_fact` defines global variables which should be avoided unless absolutely necessary. 
Since the following variables are used only inside this block please consider using `vars` section instead

miskopo replied:
@wladich Thank you for your feedback, change is implemented.

wladich reviewed (changes requested):
The new approach with ansible `crypto` module looks reaaly nice. Please see inline comment with minor issue.

wladich commented:
Testing: https://github.com/wladich/freeipa/pull/44

wladich commented:
@miskopo 
Provisioning succeeds but the test is failing: http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/44e1c9d2-97ae-11eb-b58f-fa163ebfa111/report.html

flo-renaud commented:
@miskopo the template https://github.com/freeipa/freeipa-pr-ci/blob/master/ansible/roles/machine/provision_ipaserver/templates/ui_test.conf.j2 needs to be configured differently with a ca-less server: instead of
```
#no_ca: True
```
there should be
```
no_ca: {{ caless }}
```
This way, the ui_test.conf file put in `/root/.ipa/ui_test.conf` will contain the right value for `no_ca` and the test will correctly detect if it's launched on a CAless server.

miskopo commented:
@flo-renaud @wladich  - https://github.com/miskopo/freeipa/pull/2 shows that `caless` environment is correctly recognized and installed.

flo-renaud commented on ansible/roles/machine/provision_ipaserver/tasks/main.yml:
```diff
@@ -5,6 +5,103 @@
+    pki_host: "{{ inventory_hostname_short }}"
+    pki_org: "Example Org"
+    meta_fqdn: "{{ inventory_hostname }}.ipa.test"
+    meta_domain: "{{ inventory_hostname }}.ipa.test"
```
Is `meta_domain` used somewhere? I couldn't find any occurrence.

miskopo replied:
It's being used on line 33:
```
28    - name: Generate server CSRs
29      openssl_csr:
30        path: "{{ pki_dir }}/ca1/{{ pki_host }}.csr"
31        privatekey_path: "{{ pki_dir }}/ca1/{{ pki_host }}.key"
32        subject:
33          commonName: "{{ meta_fqdn }}"
34          organizationName: "{{ pki_org }}"
```

flo-renaud commented on ansible/roles/machine/provision_ipaserver/tasks/main.yml:
```diff
@@ -5,6 +5,103 @@
+        --forwarder={{ dns_forwarder | quote }}
+        --no-pkinit
+      with_items:
+        - "{{ pki_host }}"
```
Since there is ony one item in the `with_items` loop, you can simplify and directly use `{{ pki_host }}` instead of `{{ item }}` in the above task.

flo-renaud reviewed (commented):
@miskopo 
this PR LGTM but proves that the webui test needs to be adapted (Test [report](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/a12ea8ae-987b-11eb-bb8e-fa163e861952/report.html) shows that `test_webui/test_host.py::test_host::test_negative_cert` tries to call request_cert although it's run on a caless server).

We can handle the test issue separately (there will probably be other tests not compatible with ca-less, where `if self.has_ca():` needs to be added). Finish first this PR, deploy on the runners and then we'll open tickets for the tests that need to be fixed.
I still have minor nitpicks but this is already in good shape. Thanks :)

miskopo commented:
@wladich I'd like to ask you for your final assessment. Thank you :)

wladich commented:
LGTM

flo-renaud reviewed (approved):
Also good for me. @netoarmando you can merge and deploy on the runners when you have some time. Thanks :)

netoarmando commented:
Changes deployed to all runners.
